service/mapper/transitions_test.go

```diff
@@ -0,0 +1,1233 @@
+	lastCommit := flow.StateCommitment{0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37, 0x13, 0x37}
+
+	forest := &mocks.Forest{
+		HasFunc: func(_ flow.StateCommitment) bool {
+			return true
+		},
+		TreeFunc: func(_ flow.StateCommitment) (*trie.MTrie, bool) {
+			return testTrie, true
+		},
+		PathsFunc: func(_ flow.StateCommitment) ([]ledger.Path, bool) {
+			return []ledger.Path{testPath1, testPath2, testPath3, testPath4, testPath5, testPath6}, true
+		},
+		ParentFunc: func(commit flow.StateCommitment) (flow.StateCommitment, bool) {
+			assert.Equal(t, nextCommit, commit)
+
+			return lastCommit, true
+		},
+	}
+	feed := &mocks.Feeder{
+		UpdateFunc: func() (*ledger.TrieUpdate, error) {
+			return testUpdate, nil
+		},
+	}
+
+	tr := &Transitions{
+		log: testLog,
+		cfg: Config{
+			IndexPayloads: true,
+		},
+		feed: feed,
+	}
+
+	t.Run("nominal case without match", func(t *testing.T) {
+		t.Parallel()
+
+		st := &State{
+			forest: forest,
+			registers: map[ledger.Path]*ledger.Payload{
+				testPath1: testPayload1,
+				testPath2: testPayload2,
+				testPath3: testPayload3,
+				testPath4: testPayload4,
+				testPath5: testPayload5,
+				testPath6: testPayload6,
+			},
+			status: StatusMatched,
+			next:   nextCommit,
+			last:   lastCommit,
+		}
+
+		err = tr.CollectRegisters(st)
+
+		assert.NoError(t, err)
+		assert.Equal(t, StatusCollected, st.status)
+	})
+
+	t.Run("indexing payloads disabled", func(t *testing.T) {
+		t.Parallel()
+
+		st := &State{
+			status: StatusMatched,
+		}
+
+		withoutPayloads := &Transitions{
+			cfg: Config{
+				IndexPayloads: false,
+			},
+			log: testLog,
+		}
+
+		err = withoutPayloads.CollectRegisters(st)
+
+		assert.NoError(t, err)
+		assert.Equal(t, StatusIndexed, st.status)
+	})
+
+	t.Run("handles invalid state", func(t *testing.T) {
+		t.Parallel()
+
+		invalidState := &State{
+			status: StatusEmpty,
+		}
+
+		err = tr.CollectRegisters(invalidState)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles failure to load tree from commit", func(t *testing.T) {
+		t.Parallel()
+
+		invalidForest := &mocks.Forest{
+			HasFunc: func(_ flow.StateCommitment) bool {
+				return false
+			},
+		}
+
+		st := &State{
+			forest: invalidForest,
+			registers: map[ledger.Path]*ledger.Payload{
+				testPath1: testPayload1,
+				testPath2: testPayload2,
+				testPath3: testPayload3,
+				testPath4: testPayload4,
+				testPath5: testPayload5,
+				testPath6: testPayload6,
+			},
+			status: StatusMatched,
+			next:   nextCommit,
+			last:   lastCommit,
+		}
+
+		err = tr.CollectRegisters(st)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles missing registers in map", func(t *testing.T) {
+		t.Parallel()
+
+		st := &State{
+			forest:    forest,
+			registers: map[ledger.Path]*ledger.Payload{}, // No registers, none of the paths should ever be found.
```

Comment:
That's not where we look for payloads. We look in each `tree` of the `forest`, and then copy the path & payload to the registers map for later indexing. The `registers` field is what we use to collect all path-payload pairs we want to index, across all chunks / trie updates that are in the finalized block.